File: FindMin.java
// Time Complexity O(log n)
// Space Complexity O(1)

class Solution {
    public int findMin(int[] nums) {
        int l = 0;
        int h = nums.length-1;

        if (nums.length == 0) return -1;

        while (l <= h) {
            if (nums[l] <= nums[h]) return nums[l];

            int mid = l + (h-l)/2;
            if (mid !=0 && nums[mid] < nums[mid-1]) {return nums[mid];}

            else if (nums[l] <= nums[mid]) {
               l = mid + 1;
            }
            else {
                h = mid - 1;
            }
        }
        return 999;
    }
}

File: findmin.py
#Time Complexity=O(logn)
#Space Complexity =O(1)
#// Did this code successfully run on Leetcode : Yes
#Any problem you faced while coding this :No

class Solution:
    def findMin(self, nums: List[int]) -> int:
        low,high=0,len(nums)-1
        n=len(nums)
        while low<=high:
            mid=low+(high-low)//2
            if nums[mid]<=nums[(mid-1+n)%n] and nums[mid]<=nums[(mid+1)%n]:
                return nums[mid]
            if nums[mid]>=nums[low]:
                if nums[high]>=nums[mid]:
                    high=mid-1
                else:
                    low=mid+1
            else:
                high=mid-1

        return -1
